### PR TITLE
Android button

### PR DESCRIPTION
**Pull request by roserafini**
Creating Search and Resolve Android Button

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -23,43 +23,65 @@ compile 'com.urx.android:widgets:1.0.0'
...
+#### 2. Add URX namespace within the XML layout
+
 Within the Root layout of your layout include the urx namespace.
 
 ``` xml
 xmlns:urx="http://schemas.android.com/apk/res-auto"
 ```
+
+#### 3. Add URX Widgets
+URX has 2 kinds of widgets
```
Comment (andrewlook) [inline README.md]: might be worth providing a bit of context here in plain english - why you would use one over the other, etc.

File: README.md [hunk 2]
```diff
@@ -23,43 +23,65 @@ compile 'com.urx.android:widgets:1.0.0'
...
-<com.urx.widget.URXButton
+<<com.urx.widget.PresentListViewButton
             android:id="@+id/urx_button"
             android:layout_width="wrap_content"
             android:layout_height="wrap_content"
             android:layout_centerVertical="true"
             android:layout_alignParentRight="true"
             android:text="Play Music"
-            urx:initialQuery="Beyonce"
+            urx:initialQuery="Ellie Goulding"
```
Comment (andrewlook) [inline README.md]: :+1:

File: widgets/src/main/java/com/urx/search/Util.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+import java.util.HashMap;
+
+import static com.urx.core.search.query.QueryBuilder.domain;
+import static com.urx.core.search.query.QueryBuilder.group;
+import static com.urx.core.search.query.QueryBuilder.raw;
+import static com.urx.core.search.query.QueryBuilder.term;
+
+public class Util {
+
+    public static Query createQueryFromParameters(String queryString, String actionType, String domainString){
```
Comment (andrewlook) [inline widgets/src/main/java/com/urx/search/Util.java]: minor nit: it's generally considered an antipattern to put the datatype in a variable name - this might be more appropriate:
```java
createQueryFromParameters(String keywords, String actionType, String domain)
```

File: widgets/src/main/java/com/urx/search/Util.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        }
+
+        // ActionType Query
+        /*
+        Actions are specified in the widget xml layout using comma separated values.
+        Example : urx:actionType="ListenAction,BuyAction". To know more Action visit http://schema.org/Action
+         */
+        Query actions = null;
+        if(actionType != null && !actionType.isEmpty()){
+            ArrayList<String> actionTypeList = new ArrayList<>(Arrays.asList((actionType.replaceAll("\\s+", "").split(","))));
```
Comment (andrewlook) [inline widgets/src/main/java/com/urx/search/Util.java]: phew, lots going on here.

A few notes:
* you don't need to do `new ArrayList<>(Arrays.asList...)`. Generally it's best to declare the left-hand side to use the most generic interface (in this case, `List` is more generic than `ArrayList`). So `List<String> actionTypeList = Arrays.asList(...)` will work fine.
* Also, for removing whitespace, Java's built-in [String.trim()](http://docs.oracle.com/javase/7/docs/api/java/lang/String.html#trim()) should work well for you.

File: widgets/src/main/java/com/urx/search/Util.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+            if(!actionTypeList.isEmpty()){
+                for(String actionStringType : actionTypeList){
+                    Query action = raw("action:" + actionStringType);
+                    if (actions == null) {
+                        actions = action;
+                    } else {
+                        actions = actions.or(action);
+                    }
+                }
+                queryComponents.add(group(actions));
```
Comment (andrewlook) [inline widgets/src/main/java/com/urx/search/Util.java]: nice! I like this block of code :+1:

File: widgets/src/main/java/com/urx/search/Util.java [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        }
+
+        // Query for Domain Filter
+        /*
+        Domains are specified in the widget xml layout using comma separated values.
+        Example : urx:domain="abcd.com,resetd.in,rere.org". Make sure to specify the top level domain as well.
+         */
+        Query domains = null;
+        if(domainString != null && !domainString.isEmpty()) {
+            ArrayList<String> domainList = new ArrayList<>(Arrays.asList((domainString.replaceAll("\\s+", "").split(","))));
```
Comment (andrewlook) [inline widgets/src/main/java/com/urx/search/Util.java]: same as above re:Lists and whitespace. might be worth making a helper function for "split, trim, and return a list"?

File: widgets/src/main/java/com/urx/widget/ResolveButton.java [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+    RelativeLayout rootResolveButtonLayout;
+
+    public ResolveButton(Context context, AttributeSet attrs) {
+        super(context, attrs);
+        this.context = context;
+        TypedArray typedArray = context.getTheme().obtainStyledAttributes(attrs, R.styleable.ResolveButton, 0, 0);
+        int id = typedArray.getResourceId(R.styleable.ResolveButton_buttonImage, 0);
+        if (id != 0) {
+            image_id = id;
+        }
```
Comment (andrewlook) [inline widgets/src/main/java/com/urx/widget/ResolveButton.java]: what is this block of code doing? is the assumption that if an `id` is `0`, the property is unset?

File: widgets/src/main/java/com/urx/widget/ResolveButton.java [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+        this.resolution_url = typedArray.getString(R.styleable.ResolveButton_resolution_url);
+        this.buttonText = typedArray.getString(R.styleable.ResolveButton_buttonText);
```
Comment (andrewlook) [inline widgets/src/main/java/com/urx/widget/ResolveButton.java]: what happens if these are missing?

File: widgets/src/main/java/com/urx/widget/ResolveButton.java [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+
+        updateViews();
+    }
+
+    protected void updateViews() {
+        String myApiKey = null;
+        try {
+            ApplicationInfo ai = context.getPackageManager().getApplicationInfo(context.getPackageName(), PackageManager.GET_META_DATA);
+            Bundle bundle = ai.metaData;
+            myApiKey = bundle.getString(API_KEY_STRING);
```
Comment (andrewlook) [inline widgets/src/main/java/com/urx/widget/ResolveButton.java]: should we get the API key like this every single time the page loads? or can we save it as an instance variable somewhere?

File: widgets/src/main/java/com/urx/widget/ResolveButton.java [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+        } catch (PackageManager.NameNotFoundException e) {
+            e.printStackTrace();
+        } catch (NullPointerException e) {
+            throw new NullPointerException(context.getString(R.string.set_api_key));
+        }
```
Comment (andrewlook) [inline widgets/src/main/java/com/urx/widget/ResolveButton.java]: Hmm i'm not sure what the accepted way of dealing with errors is for SDK developers. Should people expect an exception to get thrown from third party code? Is there any way that developers can catch these exceptions? or would we be better off logging them and aborting our search code but leaving their app's runtime intact?

File: widgets/src/main/java/com/urx/widget/SearchButton.java [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+import com.urx.search.Util;
+
+import static com.urx.android.resolve.AndroidResolver.installedApps;
+import static com.urx.android.resolve.AndroidResolver.playStore;
+
+
+public class SearchButton extends RelativeLayout implements View.OnClickListener {
+
+    private Context context;
+    AndroidClient client;
```
Comment (andrewlook) [inline widgets/src/main/java/com/urx/widget/SearchButton.java]: seems like the button classes have quite a lot in common - is it worth extracting a parent class? How do you think about whether this is worth it?

File: widgets/src/main/java/com/urx/widget/SearchButton.java [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+        this.client = new AndroidClient(config);
+
+        // Execute the query. Dealing with null types within createQueryFromParameters
+        final Query q = Util.createQueryFromParameters(getStringFromView(), this.actionType, this.domain);
+
+        client.query(q, new ResponseHandler<SearchResults>() {
+            @Override
+            public void onSuccess(SearchResults results) {
+                // Update Button using first result
+                searchResult = results.getResults().get(0);
```
Comment (andrewlook) [inline widgets/src/main/java/com/urx/widget/SearchButton.java]: what if there are no search results?

**Comment by andrewlook** [comment]
@roserafini overall this is looking great!

Broader feedback: there's some code duplication in the `*Button` classes, so I'd recommend considering if you can extract a parent class to encapsulate some of the common functionality (grabbing API keys, resolving a URL) and member variables (AndroidClient, Context, API_KEY_STRING).

Also, it looks like it would be worthwhile to throw in some unit tests both for the `Button` classes as well as for `Util.createQueryFromParameters`, since those a both pretty logic-heavy code paths.

File: README.md [hunk 13]
```diff
@@ -23,43 +23,65 @@ compile 'com.urx.android:widgets:1.0.0'
...
+``` xml
+
+    <com.urx.widget.SearchButton
+        android:id="@+id/urx_search_button"
+        android:layout_width="match_parent"
+        android:layout_height="70dp"
+        android:layout_alignParentBottom="true"
+        urx:initialQuery="Ellie Goulding"
+        urx:actionType="ListenAction"
+        urx:domain="spotify.com"
```
Comment (jerluc) [inline README.md]: How does one specify more than one domain?

File: README.md [hunk 14]
```diff
@@ -23,43 +23,65 @@ compile 'com.urx.android:widgets:1.0.0'
...
+
+``` xml
+
+    <com.urx.widget.SearchButton
+        android:id="@+id/urx_search_button"
+        android:layout_width="match_parent"
+        android:layout_height="70dp"
+        android:layout_alignParentBottom="true"
+        urx:initialQuery="Ellie Goulding"
+        urx:actionType="ListenAction"
```
Comment (jerluc) [inline README.md]: How does one specify more than one action?

File: README.md [hunk 15]
```diff
@@ -23,43 +23,65 @@ compile 'com.urx.android:widgets:1.0.0'
...
+        urx:domain="spotify.com"
+        urx:defaultToStore="true"/>
```
Comment (jerluc) [inline README.md]: Perhaps a more clarfiying name would be `defaultToPlayStore` or `defaultToAppStore`?

File: widgets-example/src/main/java/com/urx/example/widgets/ListButtonActivity.java [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+public class ListButtonActivity extends Activity {
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_presentlistbutton);
+    }
+
+    @Override
+    public boolean onCreateOptionsMenu(Menu menu) {
```
Comment (jerluc) [inline widgets-example/src/main/java/com/urx/example/widgets/ListButtonActivity.java]: Looks like some auto-generated cruft from using Android Studios' "New Activity" option. Is it worth removing this extra options menu code to really simplify the example code?

File: widgets-example/src/main/java/com/urx/example/widgets/ResolveButtonActivity.java [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+public class ResolveButtonActivity extends Activity {
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_resolvebutton);
+    }
+
+    @Override
+    public boolean onCreateOptionsMenu(Menu menu) {
```
Comment (jerluc) [inline widgets-example/src/main/java/com/urx/example/widgets/ResolveButtonActivity.java]: Same as my comment above regarding the `ListButtonActivity`

File: widgets-example/src/main/res/drawable-xhdpi/button_press.xml [hunk 18]
```diff
@@ -0,0 +1,27 @@
+<?xml version="1.0" encoding="utf-8"?>
+<layer-list xmlns:android="http://schemas.android.com/apk/res/android">
```
Comment (jerluc) [inline widgets-example/src/main/res/drawable-xhdpi/button_press.xml]: Are these colors skinnable by the publisher app developers? Or do we control this statically?

File: widgets/src/main/java/com/urx/search/Util.java [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+import static com.urx.core.search.query.QueryBuilder.domain;
+import static com.urx.core.search.query.QueryBuilder.group;
+import static com.urx.core.search.query.QueryBuilder.raw;
+import static com.urx.core.search.query.QueryBuilder.term;
+
+public class Util {
+
+    public static Query createQueryFromParameters(String queryString, String actionType, String domainString){
+        Query q = null;
+        ArrayList<Query> queryComponents = new ArrayList<>();
```
Comment (jerluc) [inline widgets/src/main/java/com/urx/search/Util.java]: Typically, using a more generic interface on the left is preferable to a more specific implementation (unless you are using more implementation-specific features), i.e. using `List<Query>` in this scenario. This makes any future refactoring more concise.

File: widgets/src/main/java/com/urx/search/Util.java (hunk 4 above)
Comment (jerluc) [inline widgets/src/main/java/com/urx/search/Util.java]: Same as above.

File: widgets/src/main/java/com/urx/search/Util.java [hunk 20]
```diff
@@ -0,0 +1,87 @@
...
+
+        // ActionType Query
+        /*
+        Actions are specified in the widget xml layout using comma separated values.
+        Example : urx:actionType="ListenAction,BuyAction". To know more Action visit http://schema.org/Action
+         */
+        Query actions = null;
+        if(actionType != null && !actionType.isEmpty()){
+            ArrayList<String> actionTypeList = new ArrayList<>(Arrays.asList((actionType.replaceAll("\\s+", "").split(","))));
+            if(!actionTypeList.isEmpty()){
```
Comment (jerluc) [inline widgets/src/main/java/com/urx/search/Util.java]: Under what condition would this `if` statement fail?

Comment (jerluc) [inline widgets/src/main/java/com/urx/search/Util.java]: This `if` check is unnecessary, as under such condition, the `for` loop below will just iterate zero times.

File: widgets/src/main/java/com/urx/widget/ResolveButton.java [hunk 21]
```diff
@@ -0,0 +1,104 @@
...
+    TextView descriptionTextView;
+    ImageView resolveImageView;
+    RelativeLayout rootResolveButtonLayout;
+
+    public ResolveButton(Context context, AttributeSet attrs) {
+        super(context, attrs);
+        this.context = context;
+        TypedArray typedArray = context.getTheme().obtainStyledAttributes(attrs, R.styleable.ResolveButton, 0, 0);
+        int id = typedArray.getResourceId(R.styleable.ResolveButton_buttonImage, 0);
+        if (id != 0) {
```
Comment (jerluc) [inline widgets/src/main/java/com/urx/widget/ResolveButton.java]: What should happen if the `id` does in fact equal `0`?

File: widgets/src/main/java/com/urx/widget/ResolveButton.java (hunk 10 above)
Comment (jerluc) [inline widgets/src/main/java/com/urx/widget/ResolveButton.java]: The error handling here should be cleaned up a bit.

File: widgets/src/main/java/com/urx/widget/ResolveButton.java [hunk 22]
```diff
@@ -0,0 +1,104 @@
...
+            Bundle bundle = ai.metaData;
+            myApiKey = bundle.getString(API_KEY_STRING);
+        } catch (PackageManager.NameNotFoundException e) {
+            e.printStackTrace();
+        } catch (NullPointerException e) {
+            throw new NullPointerException(context.getString(R.string.set_api_key));
+        }
+
+        if (myApiKey.equals("INSERT_KEY")) {
+            throw new RuntimeException(context.getString(R.string.set_api_key));
```
Comment (jerluc) [inline widgets/src/main/java/com/urx/widget/ResolveButton.java]: Maybe a more descriptive error message (maybe even a custom exception type?) could help here.

File: widgets/src/main/java/com/urx/widget/SearchButton.java (hunk 11 above)
Comment (jerluc) [inline widgets/src/main/java/com/urx/widget/SearchButton.java]: I don't care too much about how accessible each of these object members are, but in general, they should at least be consistent within the bounds of their usage.

File: widgets/src/main/java/com/urx/widget/SearchButton.java [hunk 23]
```diff
@@ -0,0 +1,143 @@
...
+    private boolean defaultToStore;
+    private int _id;
+    private SearchResult searchResult;
+    public static final String API_KEY_STRING = "com.urx.key";
+
+    TextView descriptionTextView;
+    ImageView searchResultImageView;
+    RelativeLayout rootSearchButtonLayout;
+
+    public SearchButton(Context context, AttributeSet attrs) {
```
Comment (jerluc) [inline widgets/src/main/java/com/urx/widget/SearchButton.java]: A large portion of the code in this class looks identical to the `ResolveButton`. Is there any way to form better abstractions such that the common code can be shared (maybe an abstract class or having the `ResolveButton` extend and override the `SearchButton`)?

**Comment by jerluc** [comment]
@roserafini overall, great job getting this code together, the results look awesome. Most of the comments I've made pertain to code style and clarity, which is a good thing! If you're available tomorrow, I'd be down to go through this code review with you in-person so we can address all of the comments from above and push this to completion.

File: widgets-example/src/main/java/com/urx/example/widgets/MainActivity.java [hunk 24]
```diff
@@ -0,0 +1,38 @@
...
+import android.view.View;
+import android.widget.Button;
+
+public class MainActivity extends Activity {
+
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_main);
+
+        Button list_button = (Button) findViewById(R.id.list_button);
```
Comment (redshirtrob) [inline widgets-example/src/main/java/com/urx/example/widgets/MainActivity.java]: Should these variable names be camel case?

cc: @aamir @jerluc

File: widgets/src/main/java/com/urx/widget/ResolveButton.java (hunk 10 above)
Comment (aamir) [inline widgets/src/main/java/com/urx/widget/ResolveButton.java]: Wanted to add here, the `Null` Pointer Exception is my doing. If folks do not have an API Key, since it will almost catch the developers attention. 

However, that being said, after talking to @redshirtrob I think this should be a feature of the SDK and not the widget, tracked [here](https://github.com/URXtech/roadmap/issues/1255)

File: README.md (hunk 13 above)
Comment (roserafini) [inline README.md]: this, along with multiple actions, is mentioned in the "Configure Widgets" section. Would you like to see it here too?

File: README.md (hunk 15 above)
Comment (roserafini) [inline README.md]: Agree and changed to defaultToPlayStore

File: widgets-example/src/main/java/com/urx/example/widgets/ListButtonActivity.java (hunk 16 above)
Comment (roserafini) [inline widgets-example/src/main/java/com/urx/example/widgets/ListButtonActivity.java]: noted and deleted

File: widgets-example/src/main/res/drawable-xhdpi/button_press.xml (hunk 18 above)
Comment (roserafini) [inline widgets-example/src/main/res/drawable-xhdpi/button_press.xml]: this is static as of rn

File: widgets/src/main/java/com/urx/widget/ResolveButton.java (hunk 7 above)
Comment (roserafini) [inline widgets/src/main/java/com/urx/widget/ResolveButton.java]: @jerluc 

For ResolveButton- results in no image, got rid of this block of code because it's handled in updateViews (this.image_id != 0)

For SearchButton- results in no view_keywordContextId, got rid of this block of code because it's handled in getStringFromView(), which returns initial query if _id is null

File: widgets/src/main/java/com/urx/widget/SearchButton.java (hunk 23 above)
Comment (roserafini) [inline widgets/src/main/java/com/urx/widget/SearchButton.java]: @andrewlook @jerluc 

Created a new abstract class URXButton which SearchButton and ResolveButton now extend.

File: widgets/src/main/java/com/urx/widget/SearchButton.java (hunk 12 above)
Comment (roserafini) [inline widgets/src/main/java/com/urx/widget/SearchButton.java]: @andrewlook 
handled in onClick() - if (searchResults != null) so the app won't crash
so basically the button will click but will not redirect.

File: widgets/src/main/java/com/urx/search/Util.java (hunk 3 above)
Comment (roserafini) [inline widgets/src/main/java/com/urx/search/Util.java]: got it

File: README.md (hunk 13 above)
Comment (jerluc) [inline README.md]: Sounds good, I'll take a look there :+1:

File: README.md (hunk 15 above)
Comment (jerluc) [inline README.md]: :+1:

File: widgets-example/src/main/java/com/urx/example/widgets/MainActivity.java (hunk 24 above)
Comment (jerluc) [inline widgets-example/src/main/java/com/urx/example/widgets/MainActivity.java]: Good point @redshirtrob. The variable name should not be (however, the `R.id.list_button` is typically snake-cased).

File: widgets-example/src/main/res/drawable-xhdpi/button_press.xml (hunk 18 above)
Comment (jerluc) [inline widgets-example/src/main/res/drawable-xhdpi/button_press.xml]: Got it :+1:
